mqtt/mqtt-broker/src/configuration.rs

```diff
@@ -0,0 +1,211 @@
+use config::{Config, ConfigError, File, FileFormat};
+use regex::Regex;
+use serde::{Deserialize, Deserializer};
+use std::path::Path;
+use std::time::Duration;
+
+pub const DEFAULTS: &str = include_str!("config.json");
+
+#[derive(Debug, Deserialize)]
+pub struct InflightMessages {
+    max_count: u32,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct RetainedMessages {
+    max_count: Option<u32>,
+    #[serde(with = "humantime_serde")]
+    expiration: Option<Duration>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct SessionMessages {
+    #[serde(deserialize_with = "humansize")]
+    max_message_size: Option<u64>,
+    max_count: Option<u32>,
+    #[serde(deserialize_with = "humansize")]
+    max_total_space: Option<u64>,
+    when_full: Option<String>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct SessionPersistence {
+    file_path: String,
+    #[serde(with = "humantime_serde")]
+    time_interval: Option<Duration>,
+    unsaved_message_count: Option<u32>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Session {
+    #[serde(with = "humantime_serde")]
+    expiration: Option<Duration>,
+    messages: SessionMessages,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct BrokerSettings {
+    inflight_messages: InflightMessages,
+    retained_messages: RetainedMessages,
+    session: Session,
+    persistence: Option<SessionPersistence>,
+}
+
+pub fn humansize<'de, D>(d: D) -> Result<Option<u64>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    lazy_static! {
+        static ref RE: Regex = Regex::new(r"\s*(\d+)\s*([a-zA-Z]+)\s*")
+            .expect("failed to create new Regex from pattern");
+    }
+
+    let s = String::deserialize(d)?;
+
+    let captures = match RE.captures(&s.as_str()) {
+        Some(s) => s,
+        None => return Err(error::<D>(&s.as_str(), &"256kb")),
+    };
+
+    let base = match captures[1].parse::<u64>() {
```

Comment:
If the regex fits, then because of the pattern, there are 2 capturing groups + the whole matching string, so "captures" should have 3 elements - if the regex library works well, and the pattern is right.